src/app-layer.c

```diff
@@ -205,11 +205,9 @@ static void TCPProtoDetectCheckBailConditions(ThreadVars *tv,
         return;
     }
 
-    const uint64_t size_ts = STREAM_HAS_SEEN_DATA(&ssn->client) ?
-        STREAM_RIGHT_EDGE(&ssn->client) : 0;
-    const uint64_t size_tc = STREAM_HAS_SEEN_DATA(&ssn->server) ?
-        STREAM_RIGHT_EDGE(&ssn->server) : 0;
-    SCLogDebug("size_ts %"PRIu64", size_tc %"PRIu64, size_ts, size_tc);
+    const uint32_t size_ts = StreamDataAvailableForProtoDetect(&ssn->client);
```

Comment:
I wonder if we still should take the diff between last_ack and isn into account here as well. Otherwise we might get all in-window data with some missing at the start. If the window is big enough this might trigger the bail conditions.

Reply:
Maybe I do not understand what you mean.
But I do not think we should take the diff between last_ack and isn into account here.
It can suddenly get really big in one packet if the TCP window is big enough (that is what oss-fuzz found).
That seems fragile, if the attacker can inject one TCP packet to evade protocol detection...
Anyways in this case, we should not use `DEBUG_VALIDATE_BUG_ON` on its value

Reply:
The scenario that I worry about is:

1. window size is bigger than the 100k we use in this function to make certain decisions
2. attacker makes sure first bytes in the stream are missing, then fills the window to the point that we reach the 100k
3. suricata gives up on protocol detect
4. attacker sends missing bytes
5. evasion

Maybe we should take the window into account, although this can be very big.

Reply:
Indeed you are right : tcp window should be used instead of hard-coded constants (but keep a minimum, and configure a maximum)

Should I add this on top of this PR ?

Reply:
Yeah please give it a try. I wonder how to deal with the potential of huge windows, as with huge windows we might use too many resources.

Reply:
Suggestion in https://github.com/OISF/suricata/pull/6189